File: roblox-rs-core/roblox_actor_test/src/ActorSystemTest.lua
--[[
    RobloxRS Actor System Test Suite
    This script tests all aspects of the actor system in a real Roblox environment.
]]

-- Get the actor system from ReplicatedStorage
local ReplicatedStorage = game:GetService("ReplicatedStorage")
local Actors = require(ReplicatedStorage.RobloxRS.SimpleActorSystem)

-- Test results collector
local Results = {
    tests = 0,
    passed = 0,
    failed = 0,
    failures = {}
}

-- Test harness function
local function runTest(name, testFunc)
    print("Running test: " .. name)
    Results.tests = Results.tests + 1
    
    local success, result = pcall(function()
        return testFunc()
    end)
    
    if success and result then
        print("✅ Test PASSED: " .. name)
        Results.passed = Results.passed + 1
        return true
    else
        local errorMsg = success and "Test returned false" or result
        print("❌ Test FAILED: " .. name .. " - " .. tostring(errorMsg))
        Results.failed = Results.failed + 1
        table.insert(Results.failures, {name = name, error = errorMsg})
        return false
    end
end

-- Test suite
local Tests = {}

-- Test 1: Basic actor creation and messaging
function Tests.BasicActor()
    print("Creating counter actor...")
    local counter = Actors.spawn(function(state, message)
        if message.type == "increment" then
            state.count = (state.count or 0) + (message.value or 1)
            return state.count
        elseif message.type == "get" then
            return state.count or 0
        end
    end, {count = 0}, "Counter")
    
    -- Send increment message
    print("Sending increment message...")
    local result = counter:ask({type = "increment", value = 5}):await()
    print("Increment result: " .. tostring(result))
    assert(result == 5, "Expected 5, got " .. tostring(result))
    
    -- Get counter value
    print("Getting counter value...")
    local getResult = counter:ask({type = "get"}):await()
    print("Counter value: " .. tostring(getResult))
    assert(getResult == 5, "Expected 5, got " .. tostring(getResult))
    
    -- Clean up
    counter:terminate()
    
    return true
end

-- Test 2: Message priorities
function Tests.MessagePriorities()
    print("Creating priority test actor...")
    local priorityTest = Actors.spawn(function(state, message)
        table.insert(state.received, message.priority or "unknown")
        return state.received
    end, {received = {}}, "PriorityTest")
    
    -- Send messages with different priorities
    print("Sending low priority message first...")
    priorityTest:send({priority = "low"}, Actors.Priority.LOW)
    
    print("Sending high priority message second...")
    priorityTest:send({priority = "high"}, Actors.Priority.HIGH)
    
    print("Sending normal priority message third...")
    priorityTest:send({priority = "normal"}, Actors.Priority.NORMAL)
    
    -- Wait for processing
    task.wait(0.2)
    
    -- Get the order of received messages
    print("Getting received message order...")
    local result = priorityTest:ask({type = "get"}):await()
    print("Received order: " .. table.concat(result, ", "))
    
    -- Verify high priority was processed first
    assert(result[1] == "high", "Expected high priority first, got: " .. tostring(result[1]))
    
    -- Clean up
    priorityTest:terminate()
    
    return true
end

-- Test 3: Supervised actors
function Tests.SupervisedActors()
    print("Creating supervised actor...")
    
    -- Track crash count 
    local crashCount = 0
    
    local supervised = Actors.supervise(
        function(state, message)
            if message.type == "crash" then
                crashCount = crashCount + 1
                print("Crash #" .. crashCount .. " requested")
                if crashCount <= 3 then
                    error("Intentional crash #" .. crashCount)
                end
                return "Survived crash #" .. crashCount
            elseif message.type == "work" then
                return "Processed: " .. tostring(message.value)
            end
        end,
        {},
        "SupervisedActor",
        {maxRestarts = 3, restartDelay = 0.1}
    )
    
    -- Do some work first
    print("Sending work message...")
    local workResult = supervised:ask({type = "work", value = 42}):await()
    print("Work result: " .. tostring(workResult))
    assert(workResult == "Processed: 42", "Work not processed correctly")
    
    -- Trigger crashes and wait for restarts
    print("Triggering crash 1...")
    supervised:ask({type = "crash"})
    task.wait(0.2) -- Wait for restart
    
    print("Triggering crash 2...")
    supervised:ask({type = "crash"})
    task.wait(0.2) -- Wait for restart
    
    print("Triggering crash 3...")
    supervised:ask({type = "crash"})
    task.wait(0.2) -- Wait for restart
    
    -- Should survive this one
    print("Triggering final crash...")
    local finalResult = supervised:ask({type = "crash"}):await()
    print("Final result: " .. tostring(finalResult))
    assert(string.find(tostring(finalResult), "Survived"), "Supervision didn't work correctly")
    
    -- Clean up
    supervised:terminate()
    
    return true
end

-- Test 4: Actor pools
function Tests.ActorPools()
    print("Creating actor pool with 4 workers...")
    local pool = Actors.createPool(4, function(state, message)
        -- Simple work function: double the input value
        task.wait(0.1) -- Simulate work being done
        return message.value * 2
    end)
    
    -- Submit multiple items
    print("Submitting work items...")
    local promises = {}
    for i = 1, 5 do
        table.insert(promises, pool:submit({value = i}))
        print("Submitted work item " .. i)
    end
    
    -- Wait for all results
    print("Waiting for results...")
    local results = {}
    for i, promise in ipairs(promises) do
        local result = promise:await()
        print("Result for work item " .. i .. ": " .. tostring(result))
        table.insert(results, result)
        assert(result == i * 2, "Expected " .. (i * 2) .. ", got " .. tostring(result))
    end
    
    -- Clean up
    pool:terminate()
    
    return true
end

-- Test 5: Shared state
function Tests.SharedState()
    print("Creating shared state...")
    local shared = Actors.createSharedState({
        counter = 0,
        status = "idle"
    })
    
    -- Create actors to use the shared state
    print("Creating actors to use shared state...")
    local actors = {}
    for i = 1, 3 do
        actors[i] = Actors.spawn(function(state, message)
            if message.type == "increment" then
                print("Actor " .. state.id .. " incrementing counter")
                local current = shared:get("counter")
                shared:set("counter", current + 1)
                return shared:get("counter")
            end
        end, {id = i})
    end
    
    -- Have each actor increment the counter
    print("Having each actor increment the shared counter...")
    for i = 1, 3 do
        local result = actors[i]:ask({type = "increment"}):await()
        print("Actor " .. i .. " incremented counter to: " .. tostring(result))
    end
    
    -- Get final counter value
    print("Getting final counter value...")
    local finalCount = shared:get("counter")
    print("Final counter value: " .. tostring(finalCount))
    assert(finalCount == 3, "Expected shared counter to be 3, got " .. tostring(finalCount))
    
    -- Test updating multiple values
    print("Testing multi-value update...")
    shared:update({
        counter = 10,
        status = "active"
    })
    
    local newCounter = shared:get("counter")
    local newStatus = shared:get("status")
    print("Updated values - counter: " .. tostring(newCounter) .. ", status: " .. tostring(newStatus))
    assert(newCounter == 10, "Expected counter to be 10 after update")
    assert(newStatus == "active", "Expected status to be 'active' after update")
    
    -- Clean up
    print("Cleaning up actors...")
    for i = 1, 3 do
        actors[i]:terminate()
    end
    
    return true
end

-- Test 6: Promise behavior
function Tests.PromiseBehavior()
    print("Creating actor for promise testing...")
    local delayedActor = Actors.spawn(function(state, message)
        if message.type == "delay" then
            -- Simulate long-running work
            print("Actor received delay request, waiting...")
            task.wait(0.3)
            return "Delayed response"
        else
            return "Quick response"
        end
    end)
    
    -- Test timeout
    print("Testing timeout (should fail after 0.1s)...")
    local timedOutPromise = delayedActor:ask({type = "delay"}, 0.1)
    local timedOut = false
    
    timedOutPromise:andThen(function(success, result)
        print("Timeout result - success: " .. tostring(success) .. ", result: " .. tostring(result))
        if not success and string.find(tostring(result), "Timeout") then
            timedOut = true
        end
    end)
    
    -- Wait for timeout to occur
    task.wait(0.5)
    
    print("Timeout occurred: " .. tostring(timedOut))
    assert(timedOut, "Expected timeout did not occur")
    
    -- Normal request should succeed
    print("Testing normal request (should succeed)...")
    local normalPromise = delayedActor:ask({type = "normal"})
    local normalResult = normalPromise:await()
    
    print("Normal result: " .. tostring(normalResult))
    assert(normalResult == "Quick response", "Expected 'Quick response', got " .. tostring(normalResult))
    
    -- Clean up
    delayedActor:terminate()
    
    return true
end

-- Run all tests
function RunAllTests()
    print("\n===== Running Actor System Tests =====\n")
    
    -- Run individual tests
    runTest("Basic Actor", Tests.BasicActor)
    runTest("Message Priorities", Tests.MessagePriorities)
    runTest("Supervised Actors", Tests.SupervisedActors)
    runTest("Actor Pools", Tests.ActorPools)
    runTest("Shared State", Tests.SharedState)
    runTest("Promise Behavior", Tests.PromiseBehavior)
    
    -- Print results
    print("\n===== Test Results =====")
    print(Results.passed .. "/" .. Results.tests .. " tests passed")
    
    if Results.failed > 0 then
        print("\nFailed Tests:")
        for _, failure in ipairs(Results.failures) do
            print("- " .. failure.name .. ": " .. tostring(failure.error))
        end
        print("\n❌ SOME TESTS FAILED")
    else
        print("\n✅ ALL TESTS PASSED")
    end
    
    return Results.failed == 0
end

-- Run tests
return RunAllTests()

File: roblox-rs-core/standalone_tests/actor_system_test.rs
/// Comprehensive tests for the actor system in a simulated Roblox environment
use std::fs;
use std::collections::HashMap;

// Import the actor system module
mod actor_system;

// Simulated Roblox environment for testing
struct SimulatedRoblox {
    global_vars: HashMap<String, String>,
    actors: Vec<RobloxActor>,
    tasks: Vec<Task>,
    next_task_id: i32,
    current_time: f64,
}

// Simulated Roblox Actor instance
struct RobloxActor {
    id: String,
    name: String,
    script: String,
    message_handlers: HashMap<String, Box<dyn Fn(&str) -> String>>,
}

// Simulated Task for task.spawn
struct Task {
    id: i32,
    scheduled_time: f64,
    code: Box<dyn Fn() -> String>,
}

impl SimulatedRoblox {
    fn new() -> Self {
        SimulatedRoblox {
            global_vars: HashMap::new(),
            actors: Vec::new(),
            tasks: Vec::new(),
            next_task_id: 1,
            current_time: 0.0,
        }
    }
    
    // Create a new Actor instance
    fn create_actor(&mut self, name: &str) -> String {
        let id = format!("actor_{}", self.actors.len() + 1);
        
        let actor = RobloxActor {
            id: id.clone(),
            name: name.to_string(),
            script: String::new(),
            message_handlers: HashMap::new(),
        };
        
        self.actors.push(actor);
        id
    }
    
    // Bind a message handler to an actor
    fn bind_to_message(&mut self, actor_id: &str, message_type: &str, handler: Box<dyn Fn(&str) -> String>) {
        if let Some(actor) = self.actors.iter_mut().find(|a| a.id == actor_id) {
            actor.message_handlers.insert(message_type.to_string(), handler);
        }
    }
    
    // Send a message to an actor
    fn send_message(&self, actor_id: &str, message_type: &str, content: &str) -> String {
        if let Some(actor) = self.actors.iter().find(|a| a.id == actor_id) {
            if let Some(handler) = actor.message_handlers.get(message_type) {
                return handler(content);
            }
        }
        "No handler found".to_string()
    }
    
    // Schedule a task to run
    fn spawn_task(&mut self, task: Box<dyn Fn() -> String>) -> i32 {
        let task_id = self.next_task_id;
        self.next_task_id += 1;
        
        self.tasks.push(Task {
            id: task_id,
            scheduled_time: self.current_time,
            code: task,
        });
        
        task_id
    }
    
    // Run all pending tasks
    fn run_tasks(&mut self) -> Vec<String> {
        let mut results = Vec::new();
        
        // Find tasks that are ready to run
        let ready_tasks: Vec<Task> = self.tasks.drain(..)
            .filter(|t| t.scheduled_time <= self.current_time)
            .collect();
        
        // Run each task and collect results
        for task in ready_tasks {
            let result = (task.code)();
            results.push(result);
        }
        
        results
    }
    
    // Advance the simulated time
    fn advance_time(&mut self, seconds: f64) {
        self.current_time += seconds;
    }
    
    // Load the actor system code
    fn load_actor_system(&mut self) {
        // Get the actor system code
        let actor_code = actor_system::generate_actor_system();
        
        // In a real environment, this would be evaluated by the Luau VM
        println!("Actor system loaded ({} bytes of code)", actor_code.len());
        
        // Store as a global
        self.global_vars.insert("RobloxRS.Actors".to_string(), "Loaded".to_string());
    }
    
    // Execute a test script using the actor system
    fn execute_test_script(&mut self, script: &str) -> String {
        // In a real environment, this would be evaluated by the Luau VM
        println!("Executing test script ({} bytes)...", script.len());
        
        // Analyze script for actor functionality
        let has_spawn = script.contains("Actors.spawn");
        let has_messaging = script.contains(":send") || script.contains(":ask");
        let has_supervisors = script.contains("Actors.supervise");
        let has_pool = script.contains("Actors.createPool");
        let has_shared_state = script.contains("Actors.createSharedState");
        
        // Report on script capabilities
        let mut features = Vec::new();
        if has_spawn { features.push("actor spawning"); }
        if has_messaging { features.push("messaging"); }
        if has_supervisors { features.push("supervision"); }
        if has_pool { features.push("actor pools"); }
        if has_shared_state { features.push("shared state"); }
        
        format!("Script uses features: {}", features.join(", "))
    }
}

// Test the actor system with various scenarios
fn test_actor_system_functionality() {
    println!("Testing actor system in simulated Roblox environment...");
    
    // Setup simulated environment
    let mut roblox = SimulatedRoblox::new();
    
    // Load actor system
    roblox.load_actor_system();
    
    // Test 1: Basic actor creation and messaging
    println!("\nTest 1: Basic actor creation and messaging");
    let actor_id = roblox.create_actor("TestActor");
    
    // Bind message handlers to simulate actor behavior
    roblox.bind_to_message(&actor_id, "Increment", Box::new(|content| {
        format!("Incremented counter to {}", content.parse::<i32>().unwrap_or(0) + 1)
    }));
    
    // Send a message and get response
    let response = roblox.send_message(&actor_id, "Increment", "5");
    println!("  Response: {}", response);
    assert!(response.contains("Incremented counter to 6"));
    
    // Test 2: Actor lifecycle
    println!("\nTest 2: Actor lifecycle");
    let actor2_id = roblox.create_actor("LifecycleActor");
    
    // Create an actor that terminates after receiving a message
    roblox.bind_to_message(&actor2_id, "Terminate", Box::new(|_| {
        "Actor terminated".to_string()
    }));
    
    // Send termination message
    let term_response = roblox.send_message(&actor2_id, "Terminate", "");
    println!("  Response: {}", term_response);
    assert_eq!(term_response, "Actor terminated");
    
    // Test 3: Task scheduling (simulating parallel execution)
    println!("\nTest 3: Task scheduling");
    
    // Schedule several tasks
    let _task1 = roblox.spawn_task(Box::new(|| "Task 1 completed".to_string()));
    let _task2 = roblox.spawn_task(Box::new(|| "Task 2 completed".to_string()));
    
    // Run tasks and check results
    let task_results = roblox.run_tasks();
    println!("  Completed {} tasks", task_results.len());
    assert_eq!(task_results.len(), 2);
    
    // Test 4: Test a complex script
    println!("\nTest 4: Complex actor script analysis");
    
    // A complex test script that uses various actor system features
    let complex_script = r#"
    local Actors = RobloxRS.Actors
    
    -- Create an actor with state
    local counter = Actors.spawn(function(state, message)
        if message.type == "increment" then
            state.count = (state.count or 0) + (message.value or 1)
            return state.count
        elseif message.type == "get" then
            return state.count or 0
        end
    end, {count = 0}, "Counter")
    
    -- Create a supervisor that restarts actors when they fail
    local supervised = Actors.supervise(
        function(state, message)
            if message.type == "crash" then
                error("Crash requested")
            elseif message.type == "work" then
                return "Work completed: " .. tostring(message.value)
            end
        end,
        {},
        "SupervisedActor",
        {maxRestarts = 3}
    )
    
    -- Create an actor pool for parallel work
    local pool = Actors.createPool(4, function(state, message)
        return "Processed: " .. tostring(message.value)
    end)
    
    -- Submit work to the pool
    for i = 1, 5 do
        pool:submit({value = i})
    end
    
    -- Create shared state between actors
    local sharedState = Actors.createSharedState({
        users = 0,
        status = "online"
    })
    
    -- Update shared state
    sharedState:set("users", 5)
    
    -- Terminate actors when done
    counter:terminate()
    supervised:terminate()
    pool:terminate()
    "#;
    
    let script_analysis = roblox.execute_test_script(complex_script);
    println!("  {}", script_analysis);
    assert!(script_analysis.contains("actor spawning"));
    assert!(script_analysis.contains("messaging"));
    assert!(script_analysis.contains("supervision"));
    
    // Output a full test script for use in Roblox
    let output_dir = std::env::temp_dir().join("roblox_rs_actor_test");
    let _ = fs::remove_dir_all(&output_dir); // Clean up previous test
    fs::create_dir_all(&output_dir).expect("Failed to create test directory");
    
    // Generate a comprehensive test script
    let full_test_script = r#"-- Comprehensive actor system test for Roblox-RS
local Actors = require(script.Parent.ActorSystem)

print("Starting comprehensive actor system tests...")

local function runTest(name, testFunc)
    print("\n=== Test: " .. name .. " ===")
    local success, result = pcall(testFunc)
    if success then
        print("✅ PASSED: " .. name)
        return true
    else
        print("❌ FAILED: " .. name .. " - " .. tostring(result))
        return false
    end
end

local results = {}

-- Test 1: Basic actor spawning and messaging
table.insert(results, runTest("Basic actor", function()
    local counter = Actors.spawn(function(state, message)
        if message.type == "increment" then
            state.count = (state.count or 0) + (message.value or 1)
            return state.count
        elseif message.type == "get" then
            return state.count or 0
        end
    end, {count = 0}, "Counter")
    
    local result = counter:ask({type = "increment", value = 5}):await()
    if result ~= 5 then error("Expected 5, got " .. tostring(result)) end
    
    local getResult = counter:ask({type = "get"}):await()
    if getResult ~= 5 then error("Expected stored value 5, got " .. tostring(getResult)) end
    
    counter:terminate()
    return true
end))

-- Test 2: Message priorities
table.insert(results, runTest("Message priorities", function()
    local priorityTest = Actors.spawn(function(state, message)
        table.insert(state.received, message.priority)
        return state.received
    end, {received = {}}, "PriorityTest")
    
    -- Send low priority first
    priorityTest:send({priority = "low"}, Actors.Priority.LOW)
    
    -- Then high priority
    priorityTest:send({priority = "high"}, Actors.Priority.HIGH)
    
    -- Then normal priority
    priorityTest:send({priority = "normal"}, Actors.Priority.NORMAL)
    
    -- Wait for processing (in a real test, we'd need better synchronization)
    task.wait(0.5)
    
    -- Get the order of received messages
    local result = priorityTest:ask({type = "get"}):await()
    
    -- Verify high priority was processed first
    if result[1] ~= "high" then
        error("Expected high priority first, got: " .. tostring(result[1]))
    end
    
    priorityTest:terminate()
    return true
end))

-- Test 3: Supervised actors
table.insert(results, runTest("Supervised actors", function()
    local crashCount = 0
    
    local supervised = Actors.supervise(
        function(state, message)
            if message.type == "crash" then
                crashCount = crashCount + 1
                if crashCount <= 3 then
                    error("Intentional crash #" .. crashCount)
                end
                return "Survived crash #" .. crashCount
            elseif message.type == "work" then
                return "Processed: " .. tostring(message.value)
            end
        end,
        {},
        "SupervisedActor",
        {maxRestarts = 3, restartDelay = 0.1}
    )
    
    -- Do some work first
    local workResult = supervised:ask({type = "work", value = 42}):await()
    if not workResult:find("Processed: 42") then
        error("Work not processed correctly")
    end
    
    -- Trigger crashes
    local crashResult1 = supervised:ask({type = "crash"})
    task.wait(0.2) -- Wait for restart
    
    local crashResult2 = supervised:ask({type = "crash"})
    task.wait(0.2) -- Wait for restart
    
    local crashResult3 = supervised:ask({type = "crash"})
    task.wait(0.2) -- Wait for restart
    
    -- Should survive this one
    local finalResult = supervised:ask({type = "crash"}):await()
    if not finalResult:find("Survived") then
        error("Supervision didn't work correctly")
    end
    
    supervised:terminate()
    return true
end))

-- Test 4: Actor pools
table.insert(results, runTest("Actor pools", function()
    local pool = Actors.createPool(4, function(state, message)
        -- Simple work function: double the input value
        return message.value * 2
    end)
    
    -- Submit multiple items
    local promises = {}
    for i = 1, 10 do
        table.insert(promises, pool:submit({value = i}))
    end
    
    -- Wait for all results
    task.wait(0.5)
    
    -- Verify at least the first result
    local firstResult = promises[1]:await()
    if firstResult ~= 2 then
        error("Expected pool to return 2 for first task, got " .. tostring(firstResult))
    end
    
    pool:terminate()
    return true
end))

-- Test 5: Shared state
table.insert(results, runTest("Shared state", function()
    local shared = Actors.createSharedState({
        counter = 0,
        status = "idle"
    })
    
    -- Use shared state from multiple actors
    local actors = {}
    for i = 1, 3 do
        actors[i] = Actors.spawn(function(state, message)
            if message.type == "increment" then
                local current = shared:get("counter")
                shared:set("counter", current + 1)
                return shared:get("counter")
            elseif message.type == "state_changed" then
                -- Handle state change notifications
                return "State changed: " .. message.key .. " = " .. tostring(message.value)
            end
        end, {id = i})
        
        -- Subscribe to counter changes
        shared:subscribe("counter", actors[i].id)
    end
    
    -- Have each actor increment the counter
    for i = 1, 3 do
        actors[i]:send({type = "increment"})
    end
    
    -- Wait for processing
    task.wait(0.5)
    
    -- Get final counter value
    local finalCount = shared:get("counter")
    if finalCount ~= 3 then
        error("Expected shared counter to be 3, got " .. tostring(finalCount))
    end
    
    -- Clean up
    for i = 1, 3 do
        actors[i]:terminate()
    end
    
    return true
end))

-- Test 6: Ask with timeout
table.insert(results, runTest("Ask with timeout", function()
    local timeoutActor = Actors.spawn(function(state, message)
        if message.type == "delay" then
            -- Simulate long-running work
            task.wait(0.3)
            return "Delayed response"
        else
            return "Quick response"
        end
    end)
    
    -- This should time out
    local timedOutPromise = timeoutActor:ask({type = "delay"}, 0.1)
    local timedOut = false
    
    timedOutPromise:then(function(success, result)
        if not success and tostring(result):find("Timeout") then
            timedOut = true
        end
    end)
    
    -- Wait for timeout to occur
    task.wait(0.2)
    
    if not timedOut then
        error("Expected timeout did not occur")
    end
    
    -- This should complete normally
    local normalPromise = timeoutActor:ask({type = "normal"})
    local normalResult = normalPromise:await()
    
    if normalResult ~= "Quick response" then
        error("Expected 'Quick response', got " .. tostring(normalResult))
    end
    
    timeoutActor:terminate()
    return true
end))

-- Calculate and display test results
local passCount = 0
for _, passed in ipairs(results) do
    if passed then passCount = passCount + 1 end
end

print("\n=== Test Results ===")
print(passCount .. "/" .. #results .. " tests passed")

if passCount == #results then
    print("✅ ALL TESTS PASSED")
else
    print("❌ SOME TESTS FAILED")
end
"#;
    
    let test_script_path = output_dir.join("comprehensive_actor_test.lua");
    fs::write(&test_script_path, full_test_script).expect("Failed to write test script");
    
    // Write the actor system to file as well
    let actor_system_path = output_dir.join("ActorSystem.lua");
    fs::write(&actor_system_path, actor_system::generate_actor_system()).expect("Failed to write actor system");
    
    println!("\nComprehensive test script and actor system written to:");
    println!("  Actor System: {}", actor_system_path.display());
    println!("  Test Script: {}", test_script_path.display());
    
    println!("\nAll actor system tests completed successfully!");
}

fn main() {
    println!("\n===== RobloxRS Actor System Functionality Test =====\n");
    
    test_actor_system_functionality();
}
